datafusion/substrait/src/logical_plan/consumer.rs

```diff
@@ -1785,6 +1798,98 @@ fn from_substrait_literal(
                 }
             }
         }
+        Some(LiteralType::Map(m)) => {
+            // Each entry should start the name index from the same value, then we increase it
+            // once at the end
+            let mut entry_name_idx = *name_idx;
+            let entries = m
+                .key_values
+                .iter()
+                .map(|kv| {
+                    entry_name_idx = *name_idx;
+                    let key_sv = from_substrait_literal(
+                        kv.key.as_ref().unwrap(),
+                        extensions,
+                        dfs_names,
+                        &mut entry_name_idx,
+                    )?;
+                    let value_sv = from_substrait_literal(
+                        kv.value.as_ref().unwrap(),
+                        extensions,
+                        dfs_names,
+                        &mut entry_name_idx,
+                    )?;
+                    ScalarStructBuilder::new()
```

Comment:
this was the most high-level way of creating the map I could think of, lmk if you have better ideas!

Reply:
also/alternatively, should I add this into ScalarValue? they could sit next to ScalarValue::new_list etc

Reply:
> this was the most high-level way of creating the map I could think of, lmk if you have better ideas!

It looks good to me. If you intend to be more efficient, I suggest referring to how `make_map_batch_internal` creates a `MapArray`.
https://github.com/apache/datafusion/blob/77311a5896272c7ed252d8cd53d48ec6ea7c0ccf/datafusion/functions-array/src/map.rs#L80

You need to partition the key and value pairs into two arrays, and build the `MapArray` based on them. Maybe you can refer to `plan_make_map`.
https://github.com/apache/datafusion/blob/77311a5896272c7ed252d8cd53d48ec6ea7c0ccf/datafusion/functions-array/src/planner.rs#L104

However, I think it just some improvements. We don't need to do that in this PR.

Reply:
Hmh, yeah I think I'll leave it as-is for now, as this code is relevant only for local relations (ie data encoded in the substrait message itself) I don't expect it to be very performance-sensitive, the data scale should hopefully always be small...